### ISSUE TITLE
Proposal - rootRef and $Auth

### ISSUE THREAD

**Issue by davideast** [firebase/angularfire#684]
## The Problem

AngularFire has no stance on where or how a developer should construct a Firebase database reference. This can lead to plenty of `new Firebase('<my-firebase-app>')` lines littered throughout a code-base. Also, to use the awesome routing helpers  like `.$waitForAuth()` and `.$requireAuth()`, you'll need to construct an `Auth` factory.

Many AngularFire apps use the following code:

```js
angular.module('app', ['firebase'])
  .constant('FirebaseUrl', '<my-firebase-app>')
  .service('rootRef', ['FirebaseUrl', Firebase])
  .factory('Auth', function ($firebaseAuth) {
     return $firebaseAuth(rootRef);
  });
```

I would love for AngularFire to handle this. 

Below are two features I propose be added to AngularFire to make authentication and constructing references easier.

### Using a provider to construct a root reference

Dependency injection in Angular means never having to say `new`. But, it's still unclear where to construct a Firebase data reference in an Angular app. AngularFire could solve this problem by using a provider. 

The provider would be responsible for getting the root URL of the Firebase database in the config phase, and then at runtime the root reference would be available for use. I propose the provider be named `rootRefProvider`, because it will then be constructed as `rootRef`.

```js
angular.module('app', ['firebase'])
   .config(function ApplicationConfig(rootRefProvider) {
      // provide the root url in the config phase
      rootRefProvider.registerRootUrl('<my-firebase-app>');
   })
   .controller('MyCtrl', function MyController(rootRef, $firebaseArray) {
      this.items = $firebaseArray(rootRef.child('items'));
   });
```

There is no `$` prefix here, because Angular's `$digest` is unaware of Firebase events. This provider simply constructs a root reference.

### $Auth

The Firebase docs on [AngularFire recommend creating an `Auth` factory](https://www.firebase.com/docs/web/libraries/angular/guide/user-auth.html#section-routers) that passes a constructed Firebase reference into the `$firebaseAuth()` service. 

With the `rootRef` service above, that could be done by AngularFire. To fit with convention, I propose it be named `$Auth`:

```js
angular.module('app', ['firebase'])
   .config(function ApplicationConfig(rootRefProvider, routeProvider) {
      // provide the root url in the config phase
      rootRefProvider.registerRootUrl('<my-firebase-app>');

      routeProvider.when('/', {
         controller: 'MyCtrl',
         template: 'view.html',
         resolve: {
           authData: function($Auth) {
              return $Auth.$requireAuth();
           }
         }
      });
   })
   .controller('MyCtrl', function($scope, authData) {
       $scope.currentUser = authData;
   });
```

@katowulf, @jamestalmage, @jwngr let me know what you guys think of this proposal. I'd love to get your input. Also, if you agree with the proposal, I would love to be the one to implement these features and also to write the tests.

I've already taken a stab at the implementation in a local branch and the code is simple. Here's a demo on JSBin.

### [Demo on JSBin](http://plnkr.co/edit/lwsGUcUGVPLcqBmcAB0z?p=preview)

**Comment by jamestalmage** [comment]
:+1: 

There are lots of benefits to this.

Following this construct makes it a lot easier to use `firebase-server`. Just swap the configured url in your tests.

Having a standardized method for the getting that configuration value and auth status seems like a good plan. One less parameter for users to pass in to reusable components (the components can just look `rootRef` themselves).

I think it is pretty rare that an app connects to two different firebases, but should you at least provide for that?

```js
rootRefProvider.registerRootUrl({
  foo: '<my-foo-app>',
  bar: '<my-bar-app>'
})
```

Finally, what about `fbRoot` instead of `rootRef`, and `fbAuth` instead of `$Auth`. The concepts of "a reference to the root of something" and "authorization status" are hardly unique to firebase.

**Comment by davideast** [comment]
Good thinking on the multiple Firebases. It's is rare, but it should be possible.

As for the names, I agree that `rootRef` and `$Auth` may hit some collisions. I'm not a huge fan of the `fb` prefix as it tends to create confusion with Facebook, who also uses `fb`.

I'll keep brainstorming on names.

**Comment by davideast** [comment]
How about `firebaseRef` and `firebaseRefProvider`?

And for registering the URL, it could be named `registerUrl()` and take in either a string or an object.

If it's a string, the `firebaseRef` will be a constructed reference.

If it's an object, the object must be a key/value of name/url. When `firebaseRef` is injected, it will be an object of keys and firebase references.

```js
angular.module('app', ['firebase'])
  .constant('FirebaseUrl', 'https://provider.firebaseio-demo.com/')
  .constant('OtherFirebaseUrl', 'https://other-firebase-db.firebaseio-demo.com/')
  .config(function(firebaseRefProvider, FirebaseUrl, OtherFirebaseUrl) {
    // register urls for several references
    firebaseRefProvider.registerUrl({
      root: FirebaseUrl,
      messages: FirebaseUrl + 'messages',
      other: OtherFirebaseUrl
    });
  })
  .controller('MyCtrl', function(firebaseRef) {
    console.log(firebaseRef.root);
    console.log(firebaseRef.messages);
    console.log(firebaseRef.other);
  });
```

### [Plnkr Demo](http://plnkr.co/edit/lwsGUcUGVPLcqBmcAB0z?p=preview)

**Comment by jamestalmage** [comment]
:+1:

**Comment by jamestalmage** [comment]
How about stealing a little from es6 module exports and allow a `default` in the object hash that is returned as `firebaseRef`, any other keys are attached as members on `firebaseRef`. There would need to be a check to make sure they did not try to overwrite an existing API method. If they don't supply a `default`, then it just returns an object hash, not a `ref` (so no namespace check needed in that case).

**Comment by davideast** [comment]
That sounds interesting, can you provide a sample of the API would look like?

**Comment by jamestalmage** [comment]
Very similar to what you have already:

```js
angular.module('app', ['firebase'])
  .constant('FirebaseUrl', 'https://provider.firebaseio-demo.com/')
  .constant('OtherFirebaseUrl', 'https://other-firebase-db.firebaseio-demo.com/')
  .config(function(firebaseRefProvider, FirebaseUrl, OtherFirebaseUrl) {
    // register urls for several references
    firebaseRefProvider.registerUrl({
      default: FirebaseUrl, // previously root
      messages: FirebaseUrl + 'messages',
      other: OtherFirebaseUrl
    });
  })
  .controller('MyCtrl', function(firebaseRef) {
    console.log(firebaseRef); // previously root
    console.log(firebaseRef.messages);
    console.log(firebaseRef.other);
  });
```

**Comment by davideast** [comment]
:+1: I like it. I'll keep thinking of names the the `fbAuth` service.

**Comment by ronnyhaase** [comment]
I was thinking about something similar while hacking around $firebaseAuth for a prototype.

But my idea was to simply turn $firebaseAuth into a provider so it's accessible in the configuration phase of Angular.
I think a constant/value for the URL, the ref, and a own service for auth/user handling is just fine.
Maybe even better than another provider that does these decisions for you?
For instance my prototype has a user service (provider) that takes care of all the $firebaseAuth internals itself - and no one else touches them! - while providing high level user functions to both, the router (which is initialized in configuration phase) and controllers.
I liked this much more than injecting the Auth factory into multiple services.

**Comment by davideast** [comment]
@ronnyhaase Sounds interesting. Could you create a Plnkr demonstrating this setup?

**Comment by davideast** [comment]
Closed with #703.
